coveragetool/branchtracer.py

```diff
@@ -0,0 +1,186 @@
+import sys
+from itertools import chain
+from collections import defaultdict
+from coveragetool.branchfinder import BranchFinder
+
+
+def func_id(function_name: str, filepath: str) -> str:
+    """Return a unique id (unique on this system!) for the given function name,
+    assuming the function name is unique in the python file at filepath.
+
+    Args:
+        function_name: Name of a function.
+        filepath: Absolute path to the file containing the function.
+    Return:
+        A system-unique id for the function.
+    """
+    return f"{function_name}@{filepath}"
+
+
+class BranchTracer:
+    """Class for tracing branch executions for select files.
+
+    Must be provided with a pre-computed nested dictionary containing mappings
+    (filepath -> (function_name -> line_nr_with_conditional)). Can be computed
+    for example with the BranchFinder class. The filepaths in this dictionary
+    are the only ones that will be traced. Tracing works only for conditional
+    statements that are contained on a single line. Multi-line statements
+    cause undefined behavior.
+
+    In addition to the constructor, the static method ``from_branchfinders`` can
+    be used to create a BranchTracer directly from a variable amount of
+    BranchFinders.
+
+    Tracing is started and stopped with the ``start_trace`` and ``stop_trace``
+    methods, which sets and unsets the global trace function using
+    ``sys.settrace``.
+    """
+
+    def __init__(self, branches: dict):
+        """
+        Args:
+            branches: A nested dictionary on the form
+            (filepath -> (function_name -> set(line_nrs_with_conditional))).
+        """
+        self._filepaths = list(branches.keys())
+
+        # mapping on the from (filepath -> set(line_nrs_with_conditional)
+        self._lines_with_condition = {
+            filepath: set(chain.from_iterable(branches.values()))
+            for filepath, branches in branches.items()
+        }
+
+        # mapping on the form
+        # (function_id -> (line_nr_with_cond -> set(lines_jumped_to)))
+        self._branches_taken = {
+            func_id(func_name, filepath): {branch: set() for branch in branches}
+            for filepath, func_dicts in branches.items()
+            for func_name, branches in func_dicts.items()
+        }
+
+        self._prev_line = dict()
+
+    def start_trace(self):
+        """Start tracing any code execution within the target files."""
+        sys.settrace(self._global_tracefunction)
+
+    def stop_trace(self):
+        """Stop tracing code execution."""
+        sys.settrace(None)
+
+    def function_coverage(self, *function_ids: str) -> dict:
+        """Get function coverage from everything traced so far as a nested
+        dictionary mapping function ids (as defined by the func_id function)
+        to dictionaries containing (line_nr_with_cond -> set of line nrs jumped
+        to) mappings.
+
+        The nested structure is thus:
+
+            (filepath ->
```

Comment:
```suggestion
            (function_id ->
```
Simple mistake I assume.

Reply:
Yeah